src/Compiler/Symbols/FSharpDiagnostic.fs

```diff
@@ -9,19 +9,104 @@ namespace FSharp.Compiler.Diagnostics
+        | ContextInfo.RuntimeTypeTest _ -> RuntimeTypeTest
+        | ContextInfo.DowncastUsedInsteadOfUpcast _ -> DowncastUsedInsteadOfUpcast
+        | ContextInfo.FollowingPatternMatchClause _ -> FollowingPatternMatchClause
+        | ContextInfo.PatternMatchGuard _ -> PatternMatchGuard
+        | ContextInfo.SequenceExpression _ -> SequenceExpression
+
+[<Interface; Experimental("This FCS API is experimental and subject to change.")>]
+type IFSharpDiagnosticExtendedData = interface end
+
+[<Experimental("This FCS API is experimental and subject to change.")>]
+type TypeMismatchDiagnosticExtendedData
+    internal (symbolEnv: SymbolEnv, dispEnv: DisplayEnv, expectedType: TType, actualType: TType, context: DiagnosticContextInfo) =
+    interface IFSharpDiagnosticExtendedData
+
+    member x.ExpectedType = FSharpType(symbolEnv, expectedType)
+    member x.ActualType = FSharpType(symbolEnv, actualType)
+    member x.ContextInfo = context
+    member x.DisplayContext = FSharpDisplayContext(fun _ -> dispEnv)
+
+[<Experimental("This FCS API is experimental and subject to change.")>]
+type ExpressionIsAFunctionExtendedData
+    internal (symbolEnv: SymbolEnv, actualType: TType) =
+    interface IFSharpDiagnosticExtendedData
+
+    member x.ActualType = FSharpType(symbolEnv, actualType)
+
+[<Experimental("This FCS API is experimental and subject to change.")>]
+type FieldNotContainedDiagnosticExtendedData
+    internal (symbolEnv: SymbolEnv, implTycon: Tycon, sigTycon: Tycon, signatureField: RecdField, implementationField: RecdField) =
+    interface IFSharpDiagnosticExtendedData
+    member x.SignatureField = FSharpField(symbolEnv, RecdFieldRef.RecdFieldRef(mkLocalTyconRef sigTycon, signatureField.Id.idText))
+    member x.ImplementationField = FSharpField(symbolEnv, RecdFieldRef.RecdFieldRef(mkLocalTyconRef implTycon, implementationField.Id.idText))
+
+[<Experimental("This FCS API is experimental and subject to change.")>]
+type ValueNotContainedDiagnosticExtendedData
+    internal (symbolEnv: SymbolEnv, signatureValue: Val, implValue: Val) =
+    interface IFSharpDiagnosticExtendedData
+    member x.SignatureValue = FSharpMemberOrFunctionOrValue(symbolEnv, mkLocalValRef signatureValue)
+    member x.ImplementationValue = FSharpMemberOrFunctionOrValue(symbolEnv, mkLocalValRef implValue)
+
+[<Experimental("This FCS API is experimental and subject to change.")>]
+type ArgumentsInSigAndImplMismatchExtendedData
+    internal(sigArg: Ident, implArg: Ident) =
+    interface IFSharpDiagnosticExtendedData
+    member x.SignatureName = sigArg.idText
+    member x.ImplementationName = implArg.idText
+    member x.SignatureRange = sigArg.idRange
+    member x.ImplementationRange = implArg.idRange
+
+type FSharpDiagnostic(m: range, severity: FSharpDiagnosticSeverity, message: string, subcategory: string, errorNum: int, numberPrefix: string, extendedData: IFSharpDiagnosticExtendedData option) =
```

Comment:
Will we only ever want to pass only singular instance of `IFSharpDiagnosticExtendedData`?

Reply:
Yes, I assume that the diagnostic will have one `ExtendedData` with all the necessary data (perhaps lazily calculable) for different quickfixes, etc.

Reply:
That's probably a good assumption to start with.